### ISSUE TITLE
Documentation

### ISSUE THREAD

**Issue by zzo38** [akuktin/special_snowflake#1]
Is there a documentation file? I could not find such a file.

**Comment by akuktin** [comment]
Hi! This project, as implemented by this codebase, has failed to deliver. I'll explain in a moment but to answer your question: there is currently no documentation. I'll whip up a small README this week and post it here. It won't really be proper documentation but at least people will be able to make heads and tails of this and make the codebase deliver a "Hello World!" of sorts.

The project is currently undergoing a near-complete redesign and rewrite and, assuming everything works, that new codebase will eventually make it's way here. Very, very tentative ETA is 2024-01-01.

Now to explain. This code **works**, but in the same sense that Cyberpunk 2077 worked when delivered. All the dots are connected, the basics are there, it "works" but it's not what people actually expect. It has two problems: doesn't close timing and bad abstraction of a key component.

The problems came to a head when I tried to put this thing into silicon. I selected iCE40 to be the chip to run it and I encountered a problem when I couldn't make the timing analysis close. At the closest approach, there was one (1) signal that was 2ns too slow. I know some people consider that to be good enough to try, but I had so much problems in my life from bad timing that I won't even try runing a design in silicon if it isn't at least 3% faster than nominal. I tried all kinds of things to make it fast enough but in the end either it's impossible, or I'm just not that good or I need better tools. At one point I actually started writing a visualizer to help me optimize but that pettered out.

This problem will be addressed in the redesign by replacing the 32-bit CPU which is currently inside, with a much more mature 16-bit CPU that already runs in silicon and has WAY more features. The reduced size should help p&r and should, hopefully, produce a fast design. I'm hoping for 66 MIPS but no promises. :)

The other problem is a bad abstraction. To put this very, very short: unlike modern computers which are organized around a bus, special_snowflake is organized around a circuit-switching switch (which I call hyperfabric). And, taking inspiration from Unix, the individual devices attached to this switch are modeled as "just streams of bits". In Linux terminology, they are character devices.

But it turns out this is a bad idea because block devices are a more general abstraction than character devices. The difference is that block devices can "rewind" (they have a fseek() of some sort), and character devices can't. You can model character devices as block devices whose "rewind buffer" has length 0, but there is no way to model block devices using character devices. Thus, things like disks are mostly impossible to interface to the hyperfabric.

This will be addressed in the redesign by the redesign. xD Devices attached to the hyperfabric are now required to conform to the block device abstraction - although they are allowed to set seek_buffer_length = 0.

So to summarize: there is no documentation, I will put some basic documentation this week and long term, this codebase will be replaced with a much better design.

**Comment by akuktin** [comment]
I finally got some time to look into this. It was bit of a challenge to find some of the old testsuites that still work. However, I'm not going to go forward with adding any documentation for the time being, for the reason that this specific codebase is being burned to the ground to make way for a new codebase, and also for the reason that then I would have to track down a working Verilog model of DDR2 memory. I have one in my development directory, from Micron, but since that package says "Copyright ... All rights reserved", I can't put it here. And tracking it down after 6 years just so I could make a little blurb that won't really help people understand a codebase that's being replaced anyway.... it just doesn't make any sense.

So I'll add a little README, which will be mostly a copy of an old post I wrote on a forum once but otherwise anything else will wait until the rewrite completes. I'm hopeful that will happen this summer.

I hope you will be patient until then. :)
